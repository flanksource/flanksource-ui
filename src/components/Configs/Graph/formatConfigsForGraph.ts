import {
  ConfigItem,
  ConfigRelationships
} from "@flanksource-ui/api/types/configs";
import { ConfigGraphNodes } from "./ConfigRelationshipGraph";

export function checkIfConfigHasMoreThan3Siblings(
  configList: Pick<
    ConfigRelationships,
    "id" | "related_id" | "direction" | "type"
  >[],
  configItem: Pick<
    ConfigRelationships,
    "id" | "related_id" | "direction" | "type"
  >
) {
  return (
    configList.filter(
      (config) =>
        config.related_id === configItem.related_id &&
        config.direction === configItem.direction
    ).length > 3
  );
}

export function shouldCreateRootIntermediaryNode(
  configList: Pick<
    ConfigRelationships,
    "id" | "related_id" | "direction" | "type"
  >[],
  configItem: Pick<
    ConfigRelationships,
    "id" | "related_id" | "direction" | "type"
  >
) {
  const types = configList
    .filter(
      (config) =>
        config.related_id === configItem.id && config.direction === "outgoing"
    )
    .reduce((acc, config) => {
      acc.add(config.type);
      return acc;
    }, new Set<string | undefined>());

  return types.size > 1;
}

export function prepareConfigsForGraph(
  configs: Pick<
    ConfigRelationships,
    "id" | "related_id" | "direction" | "type"
  >[],
  currentConfig: Pick<ConfigItem, "id" | "related_id" | "type">
) {
  const transformedConfigs: ConfigGraphNodes[] = [];

  const allConfigs = [
    ...configs,
    {
      ...currentConfig
    } as ConfigRelationships
  ];

  allConfigs.forEach((config) => {
    // if config is in the transformedConfigs, skip
    if (
      transformedConfigs.find(
        (node) => node.nodeType === "config" && node.config.id === config.id
      )
    ) {
      return;
    }

    // first, determine direct children of a config, based on direction
    // second, determine if a config has more than three children
    // if so, create an intermediary node, replace related_id with
    // intermediary node id
    const childrenConfigs = allConfigs.filter(
      (configItem) =>
        configItem.related_id === config.id &&
        configItem.direction === "outgoing"
    );

    if (childrenConfigs.length < 4) {
      // if there are less than three related configs, just push the current
      // config to the transformedConfigs
      if (!checkIfConfigHasMoreThan3Siblings(allConfigs, config)) {
        transformedConfigs.push({
          nodeType: "config",
          config: config as ConfigRelationships
        } satisfies ConfigGraphNodes);
      }
    } else {
      // When a root config can have multiple children of the same type
      // and we need to create an intermediary node to connect the children to
      // the root config
      const createRootIntermediaryNode = shouldCreateRootIntermediaryNode(
        allConfigs,
        config
      );
      // if it's a root config, we need to create an intermediary node that
      // the 3+ children will connect to
      const rootConfigIntermediaryNodeID = createRootIntermediaryNode
        ? `${config.id}-root-intermediary`
        : undefined;

      if (rootConfigIntermediaryNodeID) {
        transformedConfigs.push({
          nodeType: "intermediary",
          direction: "root" as any,
          id: rootConfigIntermediaryNodeID,
          related_id: config.id,
          configType: config.type!,
          configs: childrenConfigs as ConfigRelationships[]
        } satisfies ConfigGraphNodes);
      }

      childrenConfigs.forEach((child) => {
        const intermediaryNodeID = `${config.id}-${
          config.direction ?? "root"
        }-${child.type}`;

        // determine the number of children the config has, based on the type
        const siblingsConfigByType = allConfigs.filter(
          (configItem) =>
            configItem.related_id === config.id &&
            configItem.direction === "outgoing" &&
            configItem.type === child.type
        );

        // if the config only has one child, push the child to the
        // transformedConfigs, no need to create an intermediary node
        if (siblingsConfigByType.length > 1) {
          if (
            !transformedConfigs.find(
              (node) =>
                node.nodeType === "intermediary" &&
                node.id === intermediaryNodeID
            )
          ) {
            // This is incorrect, we need to do this based on the child configs and
            // not the parent config
            transformedConfigs.push({
              nodeType: "intermediary",
              direction: config.direction!,
              // we need to create a unique id for the intermediary node, based on
              // the config id and type
              id: intermediaryNodeID,
              // for related_id, we need to use related_id, so can place the
              // intermediary node in between the related configs
              related_id: rootConfigIntermediaryNodeID ?? config.id!,
              configType: child.type!,
              configs: siblingsConfigByType.filter(
                (c) => c.type === child.type
              ) as ConfigRelationships[]
            } satisfies ConfigGraphNodes);
          }

          // replace related_id with intermediary node id
          transformedConfigs.push({
            nodeType: "config",
            config: {
              ...(child as ConfigRelationships),
              related_id: intermediaryNodeID
            }
          });
        } else {
          // if there are less than three related configs, just push the current
          // config to the transformedConfigs
          transformedConfigs.push({
            nodeType: "config",
            config: {
              ...(child as ConfigRelationships),
              related_id: rootConfigIntermediaryNodeID ?? config.id
            }
          } satisfies ConfigGraphNodes);
        }
      });

      // Push the root config to the transformedConfigs
      if (config.id === currentConfig.id) {
        transformedConfigs.push({
          nodeType: "config",
          config: {
            ...(config as ConfigRelationships)
          }
        } satisfies ConfigGraphNodes);
      }
    }
  });
  return transformedConfigs;
}
