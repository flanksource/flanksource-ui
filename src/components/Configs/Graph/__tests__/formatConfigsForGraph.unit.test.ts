import {
  ConfigItem,
  ConfigRelationships
} from "@flanksource-ui/api/types/configs";
import { prepareConfigsForGraph } from "../formatConfigsForGraph";
import {
  applicationSetController,
  applicationSetControllerRootConfig
} from "./mocks/argo-argocd-applicationset-controller";
import {
  argoArgocdNotificationsController,
  argoArgocdNotificationsControllerRoot
} from "./mocks/argo-argocd-notifications-controller";

// Test data for ConfigRelationships
const configs: Pick<
  ConfigRelationships,
  "id" | "related_ids" | "direction" | "type"
>[] = [
  {
    id: "1",
    related_ids: ["2"],
    direction: "outgoing",
    type: "type1"
  },
  {
    id: "2",
    related_ids: ["3"],
    direction: "incoming",
    type: "type2"
  },
  {
    id: "3",
    related_ids: ["1"],
    direction: "outgoing",
    type: "type3"
  }
];

// Test data for ConfigItem
const currentConfig: Pick<ConfigItem, "id" | "related_ids" | "type"> = {
  id: "4",
  related_ids: ["1"],
  type: "type4"
};

test("prepareConfigsForGraph should return transformedConfigs", () => {
  const res = prepareConfigsForGraph(configs, currentConfig);
  expect(res).toMatchInlineSnapshot(`
    [
      {
        "config": {
          "direction": "outgoing",
          "id": "1",
          "related_ids": [
            "2",
          ],
          "type": "type1",
        },
        "nodeType": "config",
      },
      {
        "config": {
          "direction": "incoming",
          "id": "2",
          "related_ids": [
            "3",
          ],
          "type": "type2",
        },
        "nodeType": "config",
      },
      {
        "config": {
          "direction": "outgoing",
          "id": "3",
          "related_ids": [
            "1",
          ],
          "type": "type3",
        },
        "nodeType": "config",
      },
      {
        "config": {
          "id": "4",
          "related_ids": [
            "1",
          ],
          "type": "type4",
        },
        "nodeType": "config",
      },
    ]
  `);
});

test("prepareConfigsForGraph should return formatted config for argo-argocd-notifications-controller", () => {
  const res = prepareConfigsForGraph(
    argoArgocdNotificationsController,
    argoArgocdNotificationsControllerRoot
  );
  // should add an intermediary node and a root node
  expect(res.length).toEqual(argoArgocdNotificationsController.length + 2);
  expect(res).toMatchSnapshot();
});

test("prepareConfigsForGraph should return formatted config for argo-argocd-applicationset-controller-764547bfd6", () => {
  const res = prepareConfigsForGraph(
    applicationSetController,
    applicationSetControllerRootConfig
  );
  // should add an intermediary node and a root node
  expect(res.length).toEqual(argoArgocdNotificationsController.length + 2);
  expect(res).toMatchSnapshot();
});
