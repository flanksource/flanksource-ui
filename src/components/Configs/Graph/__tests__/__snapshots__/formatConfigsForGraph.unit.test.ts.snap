// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareConfigsForGraph should return formatted config for argo-argocd-applicationset-controller-764547bfd6 1`] = `
[
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "created_at": "2023-10-19T13:43:09+00:00",
      "depth": 1,
      "direction": "incoming",
      "health": "unknown",
      "id": "f5c71026-fcab-4709-ae00-2edcc7e3086a",
      "name": "argo",
      "ready": true,
      "related_ids": [
        "9580a93f-7b74-437b-836f-9a419252f3dd",
      ],
      "relation_type": "incoming",
      "status": "Healthy",
      "tags": {
        "cluster": "aws",
      },
      "type": "Kubernetes::Namespace",
      "updated_at": "2024-03-18T14:38:00.361031+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "created_at": "0001-01-01T00:00:00+00:00",
      "depth": 2,
      "direction": "incoming",
      "id": "33616137-3536-6531-6236-633631376130",
      "name": "aws",
      "ready": false,
      "related_ids": [
        "f5c71026-fcab-4709-ae00-2edcc7e3086a",
      ],
      "relation_type": "hard",
      "tags": {
        "cluster": "aws",
      },
      "type": "Kubernetes::Cluster",
      "updated_at": "2024-05-03T12:51:34.940468+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "Started",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "Scheduled",
          "severity": "",
          "total": 1,
        },
      ],
      "created_at": "2024-05-14T15:19:57+00:00",
      "depth": 2,
      "direction": "outgoing",
      "health": "healthy",
      "id": "b9317816-adbb-462e-9fde-0578b6b8d10f",
      "name": "argo-argocd-applicationset-controller-d6667d499-86bck",
      "ready": true,
      "related_ids": [
        "532fb54a-5325-431a-a618-0617020b1ab2",
      ],
      "relation_type": "hard",
      "status": "Running",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::Pod",
      "updated_at": "2024-05-14T15:20:00.574682+00:00",
    },
    "nodeType": "config",
  },
  {
    "configType": "Kubernetes::ReplicaSet",
    "configs": [
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "SuccessfulDelete",
            "severity": "",
            "total": 1,
          },
          {
            "change_type": "diff",
            "severity": "",
            "total": 4,
          },
        ],
        "created_at": "2024-05-10T08:34:05+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "unknown",
        "id": "51a1fccc-8117-4919-b249-3fb7cd8b6057",
        "name": "argo-argocd-applicationset-controller-764547bfd6",
        "ready": true,
        "related_ids": [
          "9580a93f-7b74-437b-836f-9a419252f3dd",
        ],
        "relation_type": "outgoing",
        "status": "Scaled to Zero",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-13T04:33:47.931132+00:00",
      },
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "SuccessfulDelete",
            "severity": "",
            "total": 1,
          },
          {
            "change_type": "diff",
            "severity": "",
            "total": 3,
          },
        ],
        "created_at": "2024-05-13T04:33:34+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "unknown",
        "id": "492dd2e3-aee3-45aa-91b1-a12661a6d4a8",
        "name": "argo-argocd-applicationset-controller-5cdc847b88",
        "ready": true,
        "related_ids": [
          "9580a93f-7b74-437b-836f-9a419252f3dd",
        ],
        "relation_type": "outgoing",
        "status": "Scaled to Zero",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-13T23:03:02.640177+00:00",
      },
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "created_at": "2024-05-14T15:19:57+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "healthy",
        "id": "532fb54a-5325-431a-a618-0617020b1ab2",
        "name": "argo-argocd-applicationset-controller-d6667d499",
        "ready": true,
        "related_ids": [
          "9580a93f-7b74-437b-836f-9a419252f3dd",
        ],
        "relation_type": "outgoing",
        "status": "Running",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-14T15:20:00.574682+00:00",
      },
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "SuccessfulDelete",
            "severity": "",
            "total": 1,
          },
          {
            "change_type": "diff",
            "severity": "",
            "total": 1,
          },
        ],
        "created_at": "2024-05-13T23:01:17+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "unknown",
        "id": "00f3ddcd-2f2a-4b3b-af4a-ba5091bd4501",
        "name": "argo-argocd-applicationset-controller-864b6c68",
        "ready": true,
        "related_ids": [
          "9580a93f-7b74-437b-836f-9a419252f3dd",
        ],
        "relation_type": "outgoing",
        "status": "Scaled to Zero",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-14T15:20:00.763095+00:00",
      },
    ],
    "direction": undefined,
    "id": "9580a93f-7b74-437b-836f-9a419252f3dd-root-Kubernetes::ReplicaSet",
    "nodeType": "intermediary",
    "related_id": "9580a93f-7b74-437b-836f-9a419252f3dd",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "SuccessfulDelete",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 4,
        },
      ],
      "created_at": "2024-05-10T08:34:05+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "unknown",
      "id": "51a1fccc-8117-4919-b249-3fb7cd8b6057",
      "name": "argo-argocd-applicationset-controller-764547bfd6",
      "ready": true,
      "related_ids": [
        "9580a93f-7b74-437b-836f-9a419252f3dd-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Scaled to Zero",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-13T04:33:47.931132+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "SuccessfulDelete",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 3,
        },
      ],
      "created_at": "2024-05-13T04:33:34+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "unknown",
      "id": "492dd2e3-aee3-45aa-91b1-a12661a6d4a8",
      "name": "argo-argocd-applicationset-controller-5cdc847b88",
      "ready": true,
      "related_ids": [
        "9580a93f-7b74-437b-836f-9a419252f3dd-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Scaled to Zero",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-13T23:03:02.640177+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "created_at": "2024-05-14T15:19:57+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "healthy",
      "id": "532fb54a-5325-431a-a618-0617020b1ab2",
      "name": "argo-argocd-applicationset-controller-d6667d499",
      "ready": true,
      "related_ids": [
        "9580a93f-7b74-437b-836f-9a419252f3dd-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Running",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-14T15:20:00.574682+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "SuccessfulDelete",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 1,
        },
      ],
      "created_at": "2024-05-13T23:01:17+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "unknown",
      "id": "00f3ddcd-2f2a-4b3b-af4a-ba5091bd4501",
      "name": "argo-argocd-applicationset-controller-864b6c68",
      "ready": true,
      "related_ids": [
        "9580a93f-7b74-437b-836f-9a419252f3dd-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Scaled to Zero",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-14T15:20:00.763095+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "config_class": "Deployment",
      "config_scrapers": {
        "id": "b8e78065-c9e4-40c8-bcf0-de3a39423092",
        "name": "mission-control/kubernetes",
      },
      "created_at": "2023-10-19T13:44:28+00:00",
      "delete_reason": "",
      "description": "1 pods ready",
      "external_id": [
        "9580a93f-7b74-437b-836f-9a419252f3dd",
        "Kubernetes/Deployment/argo/argo-argocd-applicationset-controller",
      ],
      "health": "healthy",
      "id": "9580a93f-7b74-437b-836f-9a419252f3dd",
      "is_pushed": true,
      "labels": {
        "apiVersion": "apps/v1",
        "app.kubernetes.io/component": "applicationset-controller",
        "app.kubernetes.io/instance": "argo",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "argocd-applicationset-controller",
        "app.kubernetes.io/part-of": "argocd",
        "app.kubernetes.io/version": "v2.11.0",
        "cluster": "aws",
        "helm.sh/chart": "argo-cd-6.9.2",
        "helm.toolkit.fluxcd.io/name": "argo",
        "helm.toolkit.fluxcd.io/namespace": "argo",
        "namespace": "argo",
      },
      "last_scraped_time": "2024-05-15T07:55:27.208541+00:00",
      "name": "argo-argocd-applicationset-controller",
      "parent_id": "f5c71026-fcab-4709-ae00-2edcc7e3086a",
      "path": "33616137-3536-6531-6236-633631376130.f5c71026-fcab-4709-ae00-2edcc7e3086a.9580a93f-7b74-437b-836f-9a419252f3dd",
      "ready": true,
      "scraper_id": "b8e78065-c9e4-40c8-bcf0-de3a39423092",
      "source": "",
      "status": "Running",
      "summary": {
        "analysis": 0,
        "changes": 0,
        "playbook_runs": 0,
        "relationships": 18,
      },
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::Deployment",
      "updated_at": "2024-05-14T15:20:00.752425+00:00",
    },
    "nodeType": "config",
  },
]
`;

exports[`prepareConfigsForGraph should return formatted config for argo-argocd-notifications-controller 1`] = `
[
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "created_at": "2023-10-19T13:43:09+00:00",
      "depth": 1,
      "direction": "incoming",
      "health": "unknown",
      "id": "f5c71026-fcab-4709-ae00-2edcc7e3086a",
      "name": "argo",
      "ready": true,
      "related_ids": [
        "b59c1752-1be4-4a51-9f84-8225ebe7c136",
      ],
      "relation_type": "incoming",
      "status": "Healthy",
      "tags": {
        "cluster": "aws",
      },
      "type": "Kubernetes::Namespace",
      "updated_at": "2024-03-18T14:38:00.361031+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "Started",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "Scheduled",
          "severity": "",
          "total": 1,
        },
      ],
      "created_at": "2024-05-14T15:19:58+00:00",
      "depth": 2,
      "direction": "outgoing",
      "health": "healthy",
      "id": "8cc44af9-61f5-452f-94c9-ab33d8cb470c",
      "name": "argo-argocd-notifications-controller-54747746f8-k966m",
      "ready": true,
      "related_ids": [
        "301902be-86af-4791-95fe-3f432c190276",
      ],
      "relation_type": "hard",
      "status": "Running",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::Pod",
      "updated_at": "2024-05-14T15:20:15.371623+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "created_at": "0001-01-01T00:00:00+00:00",
      "depth": 2,
      "direction": "incoming",
      "id": "33616137-3536-6531-6236-633631376130",
      "name": "aws",
      "ready": false,
      "related_ids": [
        "f5c71026-fcab-4709-ae00-2edcc7e3086a",
      ],
      "relation_type": "hard",
      "tags": {
        "cluster": "aws",
      },
      "type": "Kubernetes::Cluster",
      "updated_at": "2024-05-03T12:51:34.940468+00:00",
    },
    "nodeType": "config",
  },
  {
    "configType": "Kubernetes::ReplicaSet",
    "configs": [
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "SuccessfulDelete",
            "severity": "",
            "total": 1,
          },
          {
            "change_type": "diff",
            "severity": "",
            "total": 1,
          },
        ],
        "created_at": "2024-05-13T23:01:18+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "unknown",
        "id": "26a39243-80b7-4b82-941b-1ce15758b292",
        "name": "argo-argocd-notifications-controller-5b445ddccf",
        "ready": true,
        "related_ids": [
          "b59c1752-1be4-4a51-9f84-8225ebe7c136",
        ],
        "relation_type": "outgoing",
        "status": "Scaled to Zero",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-14T15:20:00.887554+00:00",
      },
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "SuccessfulDelete",
            "severity": "",
            "total": 1,
          },
          {
            "change_type": "diff",
            "severity": "",
            "total": 3,
          },
        ],
        "created_at": "2024-05-13T04:33:35+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "unknown",
        "id": "92c4c176-6570-4a5d-bdcd-b2dbc7c9a26c",
        "name": "argo-argocd-notifications-controller-777f7984cd",
        "ready": true,
        "related_ids": [
          "b59c1752-1be4-4a51-9f84-8225ebe7c136",
        ],
        "relation_type": "outgoing",
        "status": "Scaled to Zero",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-13T23:03:02.661845+00:00",
      },
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "SuccessfulDelete",
            "severity": "",
            "total": 1,
          },
          {
            "change_type": "diff",
            "severity": "",
            "total": 4,
          },
        ],
        "created_at": "2024-05-10T08:34:06+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "unknown",
        "id": "a15f91b1-dbee-4884-89af-9cd5370c2644",
        "name": "argo-argocd-notifications-controller-5f47765dfd",
        "ready": true,
        "related_ids": [
          "b59c1752-1be4-4a51-9f84-8225ebe7c136",
        ],
        "relation_type": "outgoing",
        "status": "Scaled to Zero",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-13T04:33:46.653767+00:00",
      },
      {
        "agent_id": "00000000-0000-0000-0000-000000000000",
        "changes": [
          {
            "change_type": "diff",
            "severity": "",
            "total": 1,
          },
        ],
        "created_at": "2024-05-14T15:19:58+00:00",
        "depth": 1,
        "direction": "outgoing",
        "health": "healthy",
        "id": "301902be-86af-4791-95fe-3f432c190276",
        "name": "argo-argocd-notifications-controller-54747746f8",
        "ready": true,
        "related_ids": [
          "b59c1752-1be4-4a51-9f84-8225ebe7c136",
        ],
        "relation_type": "outgoing",
        "status": "Running",
        "tags": {
          "cluster": "aws",
          "namespace": "argo",
        },
        "type": "Kubernetes::ReplicaSet",
        "updated_at": "2024-05-14T15:20:15.358938+00:00",
      },
    ],
    "direction": undefined,
    "id": "b59c1752-1be4-4a51-9f84-8225ebe7c136-root-Kubernetes::ReplicaSet",
    "nodeType": "intermediary",
    "related_id": "b59c1752-1be4-4a51-9f84-8225ebe7c136",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "SuccessfulDelete",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 1,
        },
      ],
      "created_at": "2024-05-13T23:01:18+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "unknown",
      "id": "26a39243-80b7-4b82-941b-1ce15758b292",
      "name": "argo-argocd-notifications-controller-5b445ddccf",
      "ready": true,
      "related_ids": [
        "b59c1752-1be4-4a51-9f84-8225ebe7c136-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Scaled to Zero",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-14T15:20:00.887554+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "SuccessfulDelete",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 3,
        },
      ],
      "created_at": "2024-05-13T04:33:35+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "unknown",
      "id": "92c4c176-6570-4a5d-bdcd-b2dbc7c9a26c",
      "name": "argo-argocd-notifications-controller-777f7984cd",
      "ready": true,
      "related_ids": [
        "b59c1752-1be4-4a51-9f84-8225ebe7c136-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Scaled to Zero",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-13T23:03:02.661845+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "SuccessfulDelete",
          "severity": "",
          "total": 1,
        },
        {
          "change_type": "diff",
          "severity": "",
          "total": 4,
        },
      ],
      "created_at": "2024-05-10T08:34:06+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "unknown",
      "id": "a15f91b1-dbee-4884-89af-9cd5370c2644",
      "name": "argo-argocd-notifications-controller-5f47765dfd",
      "ready": true,
      "related_ids": [
        "b59c1752-1be4-4a51-9f84-8225ebe7c136-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Scaled to Zero",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-13T04:33:46.653767+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "changes": [
        {
          "change_type": "diff",
          "severity": "",
          "total": 1,
        },
      ],
      "created_at": "2024-05-14T15:19:58+00:00",
      "depth": 1,
      "direction": "outgoing",
      "health": "healthy",
      "id": "301902be-86af-4791-95fe-3f432c190276",
      "name": "argo-argocd-notifications-controller-54747746f8",
      "ready": true,
      "related_ids": [
        "b59c1752-1be4-4a51-9f84-8225ebe7c136-root-Kubernetes::ReplicaSet",
      ],
      "relation_type": "outgoing",
      "status": "Running",
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::ReplicaSet",
      "updated_at": "2024-05-14T15:20:15.358938+00:00",
    },
    "nodeType": "config",
  },
  {
    "config": {
      "agent_id": "00000000-0000-0000-0000-000000000000",
      "config_class": "Deployment",
      "config_scrapers": {
        "id": "b8e78065-c9e4-40c8-bcf0-de3a39423092",
        "name": "mission-control/kubernetes",
      },
      "created_at": "2023-10-19T13:44:28+00:00",
      "delete_reason": "",
      "description": "1 pods ready",
      "external_id": [
        "b59c1752-1be4-4a51-9f84-8225ebe7c136",
        "Kubernetes/Deployment/argo/argo-argocd-notifications-controller",
      ],
      "health": "healthy",
      "id": "b59c1752-1be4-4a51-9f84-8225ebe7c136",
      "is_pushed": true,
      "labels": {
        "apiVersion": "apps/v1",
        "app.kubernetes.io/component": "notifications-controller",
        "app.kubernetes.io/instance": "argo",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "argocd-notifications-controller",
        "app.kubernetes.io/part-of": "argocd",
        "app.kubernetes.io/version": "v2.11.0",
        "cluster": "aws",
        "helm.sh/chart": "argo-cd-6.9.2",
        "helm.toolkit.fluxcd.io/name": "argo",
        "helm.toolkit.fluxcd.io/namespace": "argo",
        "namespace": "argo",
      },
      "last_scraped_time": "2024-05-15T07:55:27.229128+00:00",
      "name": "argo-argocd-notifications-controller",
      "parent_id": "f5c71026-fcab-4709-ae00-2edcc7e3086a",
      "path": "33616137-3536-6531-6236-633631376130.f5c71026-fcab-4709-ae00-2edcc7e3086a.b59c1752-1be4-4a51-9f84-8225ebe7c136",
      "ready": true,
      "scraper_id": "b8e78065-c9e4-40c8-bcf0-de3a39423092",
      "source": "",
      "status": "Running",
      "summary": {
        "analysis": 0,
        "changes": 0,
        "playbook_runs": 0,
        "relationships": 18,
      },
      "tags": {
        "cluster": "aws",
        "namespace": "argo",
      },
      "type": "Kubernetes::Deployment",
      "updated_at": "2024-05-14T15:20:15.364451+00:00",
    },
    "nodeType": "config",
  },
]
`;
